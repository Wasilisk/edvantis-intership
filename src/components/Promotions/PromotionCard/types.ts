import {PromotionType} from "../../../models";

export type PromotionCardProps = {
    content: PromotionType
}