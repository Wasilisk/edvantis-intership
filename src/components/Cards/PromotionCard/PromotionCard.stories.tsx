import PromotionCard from "./index";
import {Story} from "@storybook/react";
import React from "react";
import {BrowserRouter} from "react-router-dom";
import {PromotionType} from "../../../models";

const mockProductCard: PromotionType = {
    id: 1,
    image: {
        url: "https://res2.weblium.site/res/5e2f1da2c947890021c3f194/5e334f7786cda400222d1564_optimized_1396?nowebp",
        alt: "Modern Collection"
    },
    title: {
        highlight: "NEW!",
        after_highlight: "Modern Collection"
    },
    link: {
        name: "Learn more",
        to: "#promotions"
    }
}

export default {
    title: 'Application/Cards/PromotionCard',
    component: PromotionCard,
};

const Template: Story<PromotionType> = () => <PromotionCard {...mockProductCard}/>;

export const Default = Template.bind({});
Default.decorators = [
    (Story: Story) => {
        return (
            <BrowserRouter
            >
                <Story/>
            </BrowserRouter>
        );
    },
];