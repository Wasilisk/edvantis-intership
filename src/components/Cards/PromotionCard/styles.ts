import styled from "styled-components";
import {CustomLinkStyles} from "../../common/CustomLink/styles";

export const PromotionCardStyles = styled.div`
  flex-basis: 33%;
  max-width: 33%;
  padding: 0 24px;
  box-sizing: border-box;

  
  .card-image {
    display: block;
    width: 100%;
    height: 20vw;
    max-height: 250px;

    img {
      width: 100%;
      height: 100%;
      object-fit: cover;
    }
  }
  
  span {
    font-weight: 600;
  }
  
  h4 {
    margin-top: 20px;
    font-size: 22px;
    font-weight: 500;
    font-family: "Poppins", sans-serif;
    line-height: 1.3;
    color: #333333;
  }
  
  ${CustomLinkStyles} {
    margin-top: 30px;
  }

  @media (max-width: 768px) {
    flex-basis: 100%;
    max-width: 100%;
    width: 100%;
    margin-top: 50px;

    .card-image {
      height: 70vw;
      max-height: 470px;
      min-height: 184px;
    }
  }
`;