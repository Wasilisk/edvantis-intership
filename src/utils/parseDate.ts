const month = ["January","February","March","April","May","June","July","August","September","October","November","December"];

export const parseDate = (timestamp: number) => {
    const date = new Date(timestamp);

    return `${month[date.getMonth()]} ${('0' + date.getDate()).slice(-2)}, ${date.getFullYear()}`;
}