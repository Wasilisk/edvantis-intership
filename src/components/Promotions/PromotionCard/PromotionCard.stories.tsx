import PromotionCard from "./index";
import {PromotionCardProps} from "./types";
import {Story} from "@storybook/react";
import React from "react";
import {BrowserRouter} from "react-router-dom";

export default {
    title: 'Application/PromotionCard',
    component: PromotionCard,
};

const Template: Story<PromotionCardProps> = (props) => <PromotionCard content={props.content}/>;

export const Default = Template.bind({});
Default.decorators = [
    (Story: Story) => {
        return (
            <BrowserRouter
            >
                <Story/>
            </BrowserRouter>
        );
    },
];
Default.args = {
    content: {
        id: 1,
        image: {
            url: "https://res2.weblium.site/res/5e2f1da2c947890021c3f194/5e334f7786cda400222d1564_optimized_1396?nowebp",
            alt: "Modern Collection"
        },
        title: {
            highlight: "NEW!",
            plain_text: "Modern Collection"
        },
        link: {
            name: "Learn more",
            to: "#promotions"
        }
    }
};