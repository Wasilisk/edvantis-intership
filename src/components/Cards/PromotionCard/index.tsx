import React from 'react';
import {PromotionCardStyles} from "./styles";
import CustomLink from '../../common/CustomLink';
import StyledText from "../../common/StyledText";
import {PromotionType} from "../../../models";

const PromotionCard = ({image, title, link}: PromotionType) => {
    return (
        <PromotionCardStyles>
            <div className="card-image">
                <img src={image?.url} alt={image.alt}/>
            </div>
            <h4>
                <StyledText {...title}/>
            </h4>
            <CustomLink to={link?.to} label={link?.name}/>
        </PromotionCardStyles>
    );
};

export default PromotionCard;