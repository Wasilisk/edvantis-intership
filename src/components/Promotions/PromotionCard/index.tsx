import React from 'react';
import {PromotionCardProps} from "./types";
import {PromotionCardStyles} from "./styles";
import CustomLink from '../../common/CustomLink';

const PromotionCard = ({content}: PromotionCardProps) => {
    const {image, title, link} = content

    return (
        <PromotionCardStyles>
            <div className="card-image">
                <img src={image.url} alt={image.alt}/>
            </div>
            <h4>
                {
                    title.highlight && <span className="highlight">{title.highlight}&nbsp;</span>
                }
                {title.plain_text}
            </h4>
            <CustomLink to={link.to} label={link.name}/>
        </PromotionCardStyles>
    );
};

export default PromotionCard;