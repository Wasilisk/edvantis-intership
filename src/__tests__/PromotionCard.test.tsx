import {render} from "./utils/store-wrapper";
import {unstable_HistoryRouter as HistoryRouter} from "react-router-dom";
import {createMemoryHistory} from "history";
import {fireEvent} from "@testing-library/react";
import PromotionCard from "../components/Promotions/PromotionCard";
import {mockPromotions} from "./__mock__";

describe('PromotionCard component', () => {
    it('should render component', () => {
        const history = createMemoryHistory({ initialEntries: ['/'] });
        const {getByText, asFragment, getByAltText} = render(
            <HistoryRouter history={history}>
                <PromotionCard content={mockPromotions[0]}/>
            </HistoryRouter>
            );

        const cardImage = getByAltText(mockPromotions[0].image.alt);
        expect(cardImage).toHaveAttribute('src', mockPromotions[0].image.url);

        const highlightText = getByText(mockPromotions[0].title.highlight!);
        expect(highlightText).toBeDefined();
        expect(highlightText).toHaveClass(`highlight`);

        const plainText = getByText(mockPromotions[0].title.plain_text);
        expect(plainText).toBeDefined();

        const link = getByText(mockPromotions[0].link.name);
        fireEvent.click(link);
        expect(history.location.pathname).toBe(mockPromotions[0].link.to);

        expect(asFragment()).toMatchSnapshot();
    })
    it('should render component without highlight text', () => {
        const history = createMemoryHistory({ initialEntries: ['/'] });
        const {queryByText, getByText, asFragment, getByAltText} = render(
            <HistoryRouter history={history}>
                <PromotionCard content={mockPromotions[2]}/>
            </HistoryRouter>
        );

        const cardImage = getByAltText(mockPromotions[2].image.alt);
        expect(cardImage).toHaveAttribute('src', mockPromotions[0].image.url);

        const highlightText = queryByText("New!");
        expect(highlightText).not.toBeInTheDocument();

        const plainText = getByText(mockPromotions[2].title.plain_text);
        expect(plainText).toBeDefined();

        const link = getByText(mockPromotions[2].link.name);
        fireEvent.click(link);
        expect(history.location.pathname).toBe(mockPromotions[2].link.to);

        expect(asFragment()).toMatchSnapshot();
    })
})