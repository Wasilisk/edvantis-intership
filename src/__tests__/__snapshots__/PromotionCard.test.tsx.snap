// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PromotionCard component should render component 1`] = `
<DocumentFragment>
  <div
    class="sc-gsnTZi kCQKCl"
  >
    <div
      class="card-image"
    >
      <img
        alt="Modern Collection"
        src="https://res2.weblium.site/res/5e2f1da2c947890021c3f194/5e334f7786cda400222d1564_optimized_1396?nowebp"
      />
    </div>
    <h4>
      <span
        class="highlight"
      >
        NEW!Â 
      </span>
      Modern Collection
    </h4>
    <a
      class="sc-bczRLJ jZFBMR"
      href="/promotions"
    >
      Learn more 
      <svg
        height="100%"
        viewBox="0 0 36 36"
        width="100%"
        x="0px"
        xmlns="http://www.w3.org/2000/svg"
        y="0px"
      >
        <path
          class="cls-1"
          d="M34,18l-5-4v3H26v2h3v3ZM24,19H20V17h4v2Zm-6,0H14V17h4v2Zm-6,0H8V17h4v2ZM4,16a2,2,0,1,1-2,2A2,2,0,0,1,4,16Z"
          data-name="28"
          fill="currentColor"
        />
      </svg>
    </a>
  </div>
</DocumentFragment>
`;
